src/drivers/imx/sdma.c

```diff
@@ -0,0 +1,848 @@
+			    dma->plat_data.channels *
+			    sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		trace_sdma_error("SDMA: Probe failure, unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+
+	pdata = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			sizeof(*pdata));
+	if (!pdata) {
+		rfree(dma->chan);
+		dma->chan = NULL;
+		trace_sdma_error("SDMA: Probe failure, unable to allocate private data");
+		return -ENOMEM;
+	}
+	dma_set_drvdata(dma, pdata);
+
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].index = channel;
+		dma->chan[channel].dma = dma;
+	}
+
+	dma->chan[0].status = COMP_STATE_READY;
+
+	pdata->chan_pdata = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				    dma->plat_data.channels *
+				    sizeof(struct sdma_chan));
+	if (!pdata->chan_pdata) {
+		ret = -ENOMEM;
+		trace_sdma_error("SDMA: probe: out of memory");
+		goto err;
+	}
+
+	pdata->contexts = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				  dma->plat_data.channels * sizeof(struct sdma_context));
+	if (!pdata->contexts) {
+		ret = -ENOMEM;
+		trace_sdma_error("SDMA: probe: unable to allocate contexts");
+		goto err;
+	}
+
+	pdata->ccb_array = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				   dma->plat_data.channels * sizeof(struct sdma_ccb));
+	if (!pdata->ccb_array) {
+		ret = -ENOMEM;
+		trace_sdma_error("SDMA: probe: unable to allocate CCBs");
+		goto err;
+	}
+
+	ret = sdma_boot(dma);
+	if (ret < 0) {
+		trace_sdma_error("SDMA: Unable to boot");
+		goto err;
+	}
+
+	goto out;
+err:
+	if (pdata->chan_pdata)
+		rfree(pdata->chan_pdata);
+	if (pdata->contexts)
+		rfree(pdata->contexts);
+	if (pdata->ccb_array)
+		rfree(pdata->ccb_array);
+	/* Failures of allocation were treated already */
+	rfree(dma_get_drvdata(dma));
+	rfree(dma->chan);
+	dma_set_drvdata(dma, NULL);
+	dma->chan = NULL;
+out:
+	trace_sdma_error("sdma_probe done, ret = %d", ret);
+	return ret;
+}
+
+static int sdma_remove(struct dma *dma) {
+	if (!dma->chan) {
+		trace_sdma_error("SDMA: Remove called without probe, that's a noop");
+		return 0;
+	}
+
+	trace_sdma_error("sdma_remove");
+
+	/* Prevent all channels except channel 0 from running */
+	dma_reg_write(dma, SDMA_HOSTOVR, 1);
+
+	/* Stop all channels except channel 0 */
+	dma_reg_write(dma, SDMA_STOP_STAT, ~1);
+
+	/* Tear down contexts */
+	sdma_teardown_contexts(dma);
+
+	/* Shut down controller */
+	sdma_shutdown(dma);
+
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	trace_sdma_error("sdma_remove done");
+	return 0;
+}
+
+static struct dma_chan_data *sdma_channel_get(struct dma *dma, unsigned chan) {
+	struct sdma_pdata *pdata = dma_get_drvdata(dma);
+	/* Ignoring channel; let's just allocate a free channel */
+
+	trace_sdma_error("sdma_channel_get");
+	for (int i = 0; i < dma->plat_data.channels; i++) {
+		struct dma_chan_data *channel = &dma->chan[i];
+		struct sdma_chan *cdata = &pdata->chan_pdata[i];
+
```

Comment:
Declare all the variable at the beginning of function. Hope this wont trigger an compiler error for unused variable.